### ISSUE TITLE
Dll files opened by multiple processes during dotnet test

### ISSUE THREAD

**Issue by svenskmand** [coverlet-coverage/coverlet#725]
When I run `dotnet test /p:CollectCoverage=true`

I get the following exceptions that some dll files built by the project are opened by multiple processes:
```
C:\Users\CAGK\Desktop\datahub>dotnet test /p:CollectCoverage=true
Test run for C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.Tests.dll(.NETCoreApp,Version=v3.1)
Microsoft (R) Test Execution Command Line Tool Version 16.3.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.

Test Run Successful.
Total tests: 20
     Passed: 20
 Total time: 15,5588 Seconds

Calculating coverage result...
C:\Users\CAGK\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error : One or more errors occurred. (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) [C:\Users\CAGK\Desktop\datahub\Datahub.Tests\Datahub.Tests.csproj]
C:\Users\CAGK\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 59 [C:\Users\CAGK\Desktop\datahub\Datahub.Tests\Datahub.Tests.csproj]
C:\Users\CAGK\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Helpers.RetryHelper.Retry(Action action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 28 [C:\Users\CAGK\Desktop\datahub\Datahub.Tests\Datahub.Tests.csproj]
C:\Users\CAGK\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Helpers.InstrumentationHelper.RestoreOriginalModule(String module, String identifier) in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 218 [C:\Users\CAGK\Desktop\datahub\Datahub.Tests\Datahub.Tests.csproj]
C:\Users\CAGK\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Coverage.GetCoverageResult() in D:\git\coverletToRelease\src\coverlet.core\Coverage.cs:line 219 [C:\Users\CAGK\Desktop\datahub\Datahub.Tests\Datahub.Tests.csproj]
C:\Users\CAGK\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.MSbuild.Tasks.CoverageResultTask.Execute() in D:\git\coverletToRelease\src\coverlet.msbuild.tasks\CoverageResultTask.cs:line 109 [C:\Users\CAGK\Desktop\datahub\Datahub.Tests\Datahub.Tests.csproj]

C:\Users\CAGK\Desktop\datahub>Unhandled exception. System.AggregateException: One or more errors occurred. (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.) (The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.)
 ---> System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52
   --- End of inner exception stack trace ---
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 59
   at Coverlet.Core.Helpers.RetryHelper.Retry(Action action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 28
   at Coverlet.Core.Helpers.InstrumentationHelper.RestoreOriginalModules() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 239
   at Coverlet.Core.Helpers.InstrumentationHelper.<.ctor>b__3_0(Object s, EventArgs e) in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 24
   at System.AppContext.OnProcessExit()
 ---> (Inner Exception #1) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #2) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #3) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #4) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #5) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #6) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #7) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #8) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---

 ---> (Inner Exception #9) System.IO.IOException: The process cannot access the file 'C:\Users\CAGK\Desktop\datahub\Datahub.Tests\bin\Debug\netcoreapp3.1\Datahub.dll' because it is being used by another process.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Coverlet.Core.Helpers.FileSystem.Copy(String sourceFileName, String destFileName, Boolean overwrite) in D:\git\coverletToRelease\src\coverlet.core\Helpers\FileSystem.cs:line 33
   at Coverlet.Core.Helpers.InstrumentationHelper.<>c__DisplayClass10_0.<RestoreOriginalModules>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 242
   at Coverlet.Core.Helpers.RetryHelper.<>c__DisplayClass0_0.<Retry>b__0() in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 26
   at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in D:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 52<---


C:\Users\CAGK\Desktop\datahub>
```

This is my `dotnet --info`:
```
C:\Users\CAGK\Desktop\datahub>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

.NET Core SDKs installed:
  3.1.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

C:\Users\CAGK\Desktop\datahub>
```
If I remove the coverage option then the command runs fine. I do not understand why this has started happening.

**Comment by MarcoRossignoli** [comment]
Some question:

Do you have only one test project?
Are you running command on solution file?
Can you try with `dotnet test /p:CollectCoverage=true -m:1`?

**Comment by MarcoRossignoli** [comment]
Advice: if possible you should use vstest integration https://github.com/tonerdo/coverlet/blob/master/Documentation/VSTestIntegration.md

**Comment by svenskmand** [comment]
Hi Marco :)

I just run the tests from the command line. I do not use Visual Studio, I use Visual Studio Code as my editor, but I run the tests from the command line using `dotnet test /p:CollectCoverage=true /p:CoverletOutputFormat=\"lcov\" /p:CoverletOutput=\"./Datahub/lcov.info\"`

When I just run `dotnet test` it works, so it has something to do with the options `/p:CollectCoverage=true /p:CoverletOutputFormat=\"lcov\" /p:CoverletOutput=\"./Datahub/lcov.info\"`, but now it works again, so the problem seems to be intermittent :S

**Comment by MarcoRossignoli** [comment]
Intermittent nature is expected on parallel test/build try with this:

```
dotnet test /p:CollectCoverage=true /p:CoverletOutputFormat=\"lcov\" /p:CoverletOutput=\"./Datahub/lcov.info\ /nodeReuse:false -m:1" 
```

**Comment by svenskmand** [comment]
I tried `dotnet test /p:CollectCoverage=true /p:CoverletOutputFormat=\"lcov\" /p:CoverletOutput=\"./Datahub/lcov.info\" /nodeReuse:false -m:1` which works, but currently the command that did not work before, i.e. `dotnet test /p:CollectCoverage=true /p:CoverletOutputFormat=\"lcov\" /p:CoverletOutput=\"./Datahub/lcov.info\"` also works.

I is possible to pass the options `/nodeReuse:false -m:1` as properties in the `csproj` file? I have done that for the other options `/p:CollectCoverage=true /p:CoverletOutputFormat=\"lcov\" /p:CoverletOutput=\"./Datahub/lcov.info\"` like this:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <!-- ... -->
    <CollectCoverage>true</CollectCoverage>
    <CoverletOutputFormat>lcov</CoverletOutputFormat>
    <CoverletOutput>./Datahub/lcov.info</CoverletOutput>
    <ParallelizeTestCollections>false</ParallelizeTestCollections>
  </PropertyGroup>
  <!-- ... -->
</Project>
```
and I would like to do it for the `/nodeReuse:false -m:1` options also so I can just run my tests using `dotnet test`.

**Comment by svenskmand** [comment]
Also I googled a little bit about running tests in parallel when searching for solutions to the problem and I found this flag `/p:ParallelizeTestCollections=false` does that do the same as `/nodeReuse:false -m:1`?

**Comment by MarcoRossignoli** [comment]
> but currently the command that did not work before,

Yes as I said it's possible because the nature of issue is "randomic" if for some timing reason who is using that dll free the lib before restoring issue doesn't bubble up...the flag `/nodeReuse:false -m:1` said to msbuild to build and run tests sequentially and to not reuse the build hosting process. ParallelizeTestCollections seem an xunit settings to serialize tests...but the issue doesn't seem related to that, I mean when tests end nobody should use that lib anymore so it's not a problem with "tests" to me.

> and I would like to do it for the /nodeReuse:false -m:1 options also so I can just run my tests using dotnet test.

I don't know if it works as xml props https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference?view=vs-2019 try and tell me if it works, on docs is referred as command line arg.

**Comment by svenskmand** [comment]
The `-m` (`--maxcpucount`) should be the property `MSBuildNodeCount` according to [the link](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-reserved-and-well-known-properties?view=vs-2019) and I think that `nodeReuse` should be the property `NodeReuse` according to [the link](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/hh727536(v%3Dvs.120)).

I will try and keep these two set and see if the issue occurs again. You think the issue occurs in the `dotnet test` command and not in coverlet?

**Comment by MarcoRossignoli** [comment]
> You think the issue occurs in the dotnet test command and not in coverlet?

It occour in coverlet but it's related to how build/test run. Coverlet restore all instrumented libs at the end of tests, we use an msbuild task(in this case) so seem that also if task should start at the end of tests someone(msbuild/test process) has got that dll loaded and doesn't allow to coverlet to restore(copy) files back.

You could try with vstest integration, maybe solve the issue because the "worflow" is completely different https://github.com/tonerdo/coverlet/blob/master/Documentation/VSTestIntegration.md and there is no msbuild interaction.

**Comment by Cheesebaron** [comment]
We have this issue as well. It happens intermittenly in Azure DevOps. However, when I run locally it happens _every_ time.

I tried adding both `-m:1` and `/nodeReuse:false` but that doesn't seem to help.

The command I am running looks something like:

```
dotnet test "/Users/cheesebaron/git/DrivingRangeApp/TrackMan.DrivingRangeApp.Tests/TrackMan.DrivingRangeApp.Tests.csproj" --settings "/Users/cheesebaron/git/DrivingRangeApp/CodeCoverage.runsettings" --logger "xunit;LogFilePath=/Users/cheesebaron/git/DrivingRangeApp/artifacts/Tests/TrackMan.DrivingRangeApp.Tests.xml" --configuration Labs --no-build --collect:"XPlat Code Coverage" -m:1 /nodeReuse:false
```

Tried with `dotnet vstest` but it can't seem to find the data collector, even though I specify it and give it the same CodeCoverage.runsettings file as `dotnet test` runs with.

**Comment by MarcoRossignoli** [comment]
It's possible have a repro?Are one of these project OSS?I remember another user with same issue but I cannot repro in my local so it's hard to understand.

Can you try with my custom patched version? https://github.com/tonerdo/coverlet/issues/572#issuecomment-539871519

**Comment by Cheesebaron** [comment]
@MarcoRossignoli it is unfortunately not OSS

**Comment by Cheesebaron** [comment]
OK! Just tried something. Lowered the version of `Microsoft.NET.Test.Sdk` from 16.5.0 to 16.4.0 and now it doesn't fail _every_ time on on my machine anymore. Will investigate a little more.

**Comment by MarcoRossignoli** [comment]
> OK! Just tried something. Lowered the version of Microsoft.NET.Test.Sdk from 16.5.0 to 16.4.0 and now it doesn't fail every time on on my machine anymore. 

I've got a suspect...can you try with my patched version? https://github.com/tonerdo/coverlet/issues/572#issuecomment-539871519

**Comment by Cheesebaron** [comment]
@MarcoRossignoli yep, on it now.

**Comment by MarcoRossignoli** [comment]
Please try removing InProc sections on xml and add `--diag:log.txt` and re-run, check that on log file *host* there is some row with [coverlet] prefix, I want to understand of you're loading inproc collector correctly

**Comment by Cheesebaron** [comment]
Yes, the host log file contains this:

```
TpTrace Warning: 0 : 5736, 4, 2020/02/12, 16:07:04.488, 3170029447655724, testhost.dll, [coverlet]CoverletInProcDataCollector: Failed to get Instrumentation class with error: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis.Workspaces, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis.Workspaces, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis.Workspaces, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

Could not load file or assembly 'Microsoft.CodeAnalysis.Workspaces, Version=2.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.

   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.RuntimeModule.GetTypes()
   at System.Reflection.Assembly.GetTypes()
   at Coverlet.Collector.DataCollection.CoverletInProcDataCollector.GetInstrumentationClass(Assembly assembly) in C:\git\coverlet\src\coverlet.collector\InProcDataCollection\CoverletInProcDataCollector.cs:line 65
```

**Comment by MarcoRossignoli** [comment]
Seem ok...the errors are [logged](https://github.com/tonerdo/coverlet/blob/master/src/coverlet.collector/InProcDataCollection/CoverletInProcDataCollector.cs#L78) so are not a problem, the essential part is 
```
TpTrace Verbose: 0 : 5736, 4, 2020/02/12, 16:07:04.478, 3170029436956824, testhost.dll, [coverlet]Calling ModuleTrackerTemplate.UnloadModule for 'TrackMan.DrivingRangeApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
TpTrace Verbose: 0 : 5736, 4, 2020/02/12, 16:07:04.481, 3170029440183782, testhost.dll, [coverlet]Called ModuleTrackerTemplate.UnloadModule for 'TrackMan.DrivingRangeApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'
```

Double check....if you only update to official 1.2.0 does it work or start to fail?

**Comment by qdraw** [comment]
I have the same issue on a Azure Devops pipeline. (vmImage: windows-2019)

Content of test project csproj
```
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <RuntimeFrameworkVersion>3.1.3</RuntimeFrameworkVersion>
```
and
```
    <PackageReference Include="coverlet.msbuild" Version="2.8.1" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
    <PackageReference Include="MSTest.TestAdapter" Version="2.1.1" />
    <PackageReference Include="MSTest.TestFramework" Version="2.1.1" />
```

```
Test Run Successful.
Total tests: 682
     Passed: 682
 Total time: 22.2975 Seconds

Calculating coverage result...
C:\Users\VssAdministrator\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error : Unable to read beyond the end of the stream. [D:\a\1\s\starsky\starskytest\starskytest.csproj]
C:\Users\VssAdministrator\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at System.IO.BinaryReader.InternalRead(Int32 numBytes) [D:\a\1\s\starsky\starskytest\starskytest.csproj]
C:\Users\VssAdministrator\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Coverage.CalculateCoverage() in c:\git\coverletToRelease\src\coverlet.core\Coverage.cs:line 355 [D:\a\1\s\starsky\starskytest\starskytest.csproj]
C:\Users\VssAdministrator\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Coverage.GetCoverageResult() in c:\git\coverletToRelease\src\coverlet.core\Coverage.cs:line 137 [D:\a\1\s\starsky\starskytest\starskytest.csproj]
C:\Users\VssAdministrator\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.MSbuild.Tasks.CoverageResultTask.Execute() in c:\git\coverletToRelease\src\coverlet.msbuild.tasks\CoverageResultTask.cs:line 113 [D:\a\1\s\starsky\starskytest\starskytest.csproj]
An error occurred when executing task 'TestNetCore'.
Error: One or more errors occurred. (.NET Core CLI: Process returned an error (exit code 1).)
	.NET Core CLI: Process returned an error (exit code 1).
```

I'm using the following configuration to execute the test project (Cake)

```
            DotNetCoreTest(
                project.ToString(),
                new DotNetCoreTestSettings()
                {
                    Configuration = configuration,
                    NoBuild = true,
                    ArgumentCustomization = args => args
                          .Append("--no-restore")
                          .Append("--no-build")
                          .Append("--nologo")
                          .Append("-m:1")
                          .Append("/p:CollectCoverage=true")
                          .Append("/p:CoverletOutputFormat=opencover")
                          .Append("/p:ThresholdType=line")
                });
```

I have tried the suggestion to append `-m:1`. This issue doesn't appear always on some builds it fails and on my local machine I don't have this issue. 

Does anyone have a suggestion what I could try?

**Comment by MarcoRossignoli** [comment]
@Cheesebaron can you try with our nightly build we fixed the issue of race condition in collectors https://github.com/coverlet-coverage/coverlet/pull/835
This is the guide of nightly https://github.com/coverlet-coverage/coverlet/blob/master/Documentation/ConsumeNightlyBuild.md

@svenskmand does this issue solved?

**Comment by Cheesebaron** [comment]
@MarcoRossignoli yes. Will have a look tomorrow. Do you want me to test with coverlet nightly _and_ latest version of Test.Sdk?

**Comment by MarcoRossignoli** [comment]
> @MarcoRossignoli yes. Will have a look tomorrow. Do you want me to test with coverlet nightly and latest version of Test.Sdk?

Yes thanks

**Comment by Cheesebaron** [comment]
It seems to work. Tried a couple of runs with coverlet.collector version 1.3.0-preview.17.g3c55332cce and Microsoft.NET.Test.Sdk 16.6.1

**Comment by MarcoRossignoli** [comment]
Great!Thanks!

**Comment by AntiPasha** [comment]
Hi, I've got the same errors, mentioned above, with both msbuild and vstest integrations. Interesting point here, the error is observed with tests written with xUnit, with NUnit everything work just fine. Maybe it's related to xunit.runner?

**Comment by MarcoRossignoli** [comment]
Can you attach the error?

**Comment by AntiPasha** [comment]
Hi @MarcoRossignoli please have a look at logs generated by MSBuild integration (I've just removed the sensitive names)
```C:\Users\xxx\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error : One or more errors occurred. (The process cannot access the file 'D:\Work\xxx\yyy\test\yyy.Tests\bin\Debug\netcoreapp2.1\zzz.Dto.dll' because it is being used by another process.)[D:\Work\xxx\yyy\test\yyy.Tests\yyy.Tests.csproj]
C:\Users\xxx\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Helpers.RetryHelper.Do[T](Func`1 action, Func`1 backoffStrategy, Int32 maxAttemptCount) in c:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 59 [D:\Work\xxx\yyy\test\yyy.Tests\yyy.Tests.csproj]
C:\Users\xxx\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Helpers.RetryHelper.Retry(Action action, Func`1 backoffStrategy, Int32 maxAttemptCount) in c:\git\coverletToRelease\src\coverlet.core\Helpers\RetryHelper.cs:line 28 [D:\Work\xxx\yyy\test\yyy.Tests\yyy.Tests.csproj]
C:\Users\xxx\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Helpers.InstrumentationHelper.RestoreOriginalModule(String module, String identifier) in c:\git\coverletToRelease\src\coverlet.core\Helpers\InstrumentationHelper.cs:line 219 [D:\Work\xxx\yyy\test\yyy.Tests\yyy.Tests.csproj]
C:\Users\xxx\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Coverage.GetCoverageResult() in c:\git\coverletToRelease\src\coverlet.core\Coverage.cs:line 219 [D:\Work\xxx\yyy\test\yyy.Tests\yyy.Tests.csproj]
C:\Users\xxx\.nuget\packages\coverlet.msbuild\2.8.1\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.MSbuild.Tasks.CoverageResultTask.Execute() in c:\git\coverletToRelease\src\coverlet.msbuild.tasks\CoverageResultTask.cs:line 113 [D:\Work\xxx\yyy\test\yyy.Tests\yyy.Tests.csproj]
```

**Comment by MarcoRossignoli** [comment]
If you run for only one csproj does this error happen?I mean run something like `dotnet test ....TestProject.csproj /p:CollectCoverage=true`

**Comment by AntiPasha** [comment]
Yes, error the same. I have only one test project here. I've tried to run using different ways, by specifying `csproj`, by running `dotnet test` at solution and project folders. Interesting part here is that it fails even I specify `--no-build` switch as well (in one of linked issues I found that instrumentation will not run in this case, maybe I'm wrong)

**Comment by MarcoRossignoli** [comment]
`--no-build` is not an issue...the error I see is the rollback of instrumentation, when tests end coverlet get a file with hits(from tmp) and calculate coverage, after that rollback dll to old ones(moved before IL instrumentation to another tmp folder so we can run more than one instrumentation and also we maintain dll in a original state), and seem that someone is locking files in your `bin/Debug` folder.
I don't think it's related to coverlet if it fails also with single project(tests was to understand if that dll was loaded more than one time in different tests, out of coverlet control but maybe could be the culprit), you could try to understand who is blocking that files(antivirus?), you can start with this guide https://support.qlik.com/articles/000005295 
I cannot repro in my local, in past we had some issue with hits file but was related to our bug with VSTest integration fixed on master(we'll release asap) https://github.com/coverlet-coverage/coverlet/pull/835 but no with dll.

**Comment by AntiPasha** [comment]
Thank you. will have a look. Can xunit.runner or testhost lock these files? Because I have a bunch of nunit tests in different repo, they work fine. Output folder for nunit tests doesn't contain testhost.exe and dll, but for xunit does

**Comment by MarcoRossignoli** [comment]
> Can xunit.runner or testhost lock these files?

I know a bit testhost and never seen that issue, I don't know nothing about xunit runner...btw it's hard to guess...could be everything, we need to analyze with some tecnique above.

**Comment by AntiPasha** [comment]
I've collected some events using Process Monitor, there no processes, except `dotnet.exe` and `testhost.exe` accessing the file. During usual test run (without coverage) there is only file open, query information and locking file with only readers by `testhost.exe`. 
When coverage is run, there are a file reading, writing and locking file with only readers by **both** processes. After file is locked by `testhost.exe`,  `dotnet.exe` is returning a sharing violation result when attempt to use a file for generic read and write. It seems that both processes have something a race condition, when file is locked for reading by one process and second process attempts to access it with read/write access

**Comment by MarcoRossignoli** [comment]
@AntiPasha two thing, first one, can you file a new issue and attach your first and last comment?This issue is old aged and stale and I want a clean history for your one.
Second, is the behaviour the same with version `2.8.1` or `1.2.0`?I mean it's first time you use coverlet or it's a version regression?

**Comment by AntiPasha** [comment]
@MarcoRossignoli Thank you, I've created new [857](https://github.com/coverlet-coverage/coverlet/issues/857) issue fro that. It's first time, when I try to use coverlet

**Comment by hippieZhou** [comment]
> Some question:
> 
> Do you have only one test project? Are you running command on solution file? Can you try with `dotnet test /p:CollectCoverage=true /nodeReuse:false -m:1`?
> 
> From exception seem that after test run when coverlet tries to "restore" old files(because coverlet to work backup file instrument and at the end restore old one without instrumentation code inside) that files are in use by someone, so restoring fail. I found this issue sometimes when lib for some reason are "re-used" for different tests(related to some build config) in parallel or when you run big solutions with a lot of test projects(by default will run in parallel) and slow tests.

I also have this problem in my multiple test projects, when I use `dotnet test -m:1`  by CLI, all my test cases can passed.

**Comment by svenskmand** [comment]
> @Cheesebaron can you try with our nightly build we fixed the issue of race condition in collectors #835 This is the guide of nightly https://github.com/coverlet-coverage/coverlet/blob/master/Documentation/ConsumeNightlyBuild.md
> 
> @svenskmand does this issue solved?

I have not had the issue in a long time, I am currently using version 3.2.0 of both coverlet.collector and coverlet.msbuild. So it might have been resolved :S

Thanks for all your help and time :)

**Comment by Cheesebaron** [comment]
Yeah, I haven't seen it forever after updating, so I think all is good. Still running coverlet so should be fine to keep this closed.

Thanks for following up :)
